firefox-ios/Client/Frontend/Browser/BrowserViewController/Extensions/BrowserViewController+WebViewDelegates.swift

```diff
@@ -1043,19 +1043,25 @@ private extension BrowserViewController {
     func handleServerTrust(challenge: URLAuthenticationChallenge,
                            completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void
     ) {
-        // If this is a certificate challenge, see if the certificate has previously been
-        // accepted by the user.
-        let origin = "\(challenge.protectionSpace.host):\(challenge.protectionSpace.port)"
+        DispatchQueue.global(qos: .userInitiated).async {
+            // If this is a certificate challenge, see if the certificate has previously been
+            // accepted by the user.
+            let origin = "\(challenge.protectionSpace.host):\(challenge.protectionSpace.port)"
+
+            guard let trust = challenge.protectionSpace.serverTrust,
+                  let cert = SecTrustCopyCertificateChain(trust) as? [SecCertificate],
+                  self.profile.certStore.containsCertificate(cert[0], forOrigin: origin)
+            else {
+                DispatchQueue.main.async {
+                    completionHandler(.performDefaultHandling, nil)
+                }
+                return
+            }
 
-        guard let trust = challenge.protectionSpace.serverTrust,
-              let cert = SecTrustCopyCertificateChain(trust) as? [SecCertificate],
-              profile.certStore.containsCertificate(cert[0], forOrigin: origin)
-        else {
-            completionHandler(.performDefaultHandling, nil)
-            return
+            DispatchQueue.main.async {
+                completionHandler(.useCredential, URLCredential(trust: trust))
+            }
```

Comment:
Q. main is still run on the main thread but async, did you mean to run this on a global dispatch instead?

Reply:
the intention was to have the completion handler get back on the main thread after the processing inside the `DispatchQueue.global(qos: .userInitiated).async` happens to avoid potential UI hangs. Global queue is a concurrent queue that is managed by the system, which means that it can run tasks on any available thread. By default, tasks that are submitted to a global queue will be executed on a non-main thread